New Vector_sort function, replacing insertion and quick sort

Explorer09 opened the pull request:
Introduce the new `Vector_sort()` function and obsolete the old `Vector_quickSortCustomCompare()` and `Vector_insertionSort()` APIs.

This new sort function is a natural, in-place merge sort. I.e. it takes advantage of partially sorted data, and it's stable.

Space complexity: O(log(n)) worst case
Time complexity: O(n) best case, O(n*log(n)*log(n)) average & worst case

@BenBE , I have copied some of the changes from your d3cd557e0534a45ab971397d4c407116547d3f6d commit but I didn't base on it. Because there are some data type naming that I didn't like, and so I name my "context" structure in my own way. Also to keep changes small I didn't upgrade the `Object_Compare` function type to take a new "context" parameter.

Explorer09 commented on Vector.c:
```diff
@@ -93,91 +101,168 @@ void Vector_prune(Vector* this) {
-         swap(array, i, storeIndex);
-         storeIndex++;
+ATTR_NONNULL
+static void rotate(void* buffer, size_t leftSize, size_t rightSize) {
```
A note to leave to myself:
There are two possible interfaces for this `rotate` function. To minimize the chances that the callers pass  invalid values, I picked the sizes as the second and third parameters. When it comes to optimization, this would be more dependent on the compiler's capability in simplifying additions and subtraction than the version that passes `mid` and `end` pointers. I didn't check the code sizes produced yet, so this remains a TODO for me.

```c
rotate_a(void* buffer, size_t leftSize, size_t rightSize);
rotate_b(void* start, void* mid, const void* end);
```

natoscott commented:
Sort stability is about what happens when two items compare equally (which can happen with/without multiple sort keys, so I don't follow your argument there).   In htop, we always have the PID we can use as a tie-breaker to guarantee stability.  Can you explain how this is insufficient?

Regarding threads and the extra parameter, htop is single threaded and is very likely to remain so.  So again this is not a valid technical argument against the use of qsort here.

qsort is highly optimised and widely reviewed.   Any custom home-grown sorting code we have in htop simply isn't, to anywhere near the same degree, and appears to be unnecessary code bloat.  So unless there is a valid technical reason otherwise (and none have been put forward so far, AFAICT), I think we should switch to qsort and delete all this code.

Explorer09 commented:
@natoscott I assume you have read #1784 and #1785 for the discussion I had with @BenBE before.

Basically I would vote for using qsort() (#1784) if there isn't a stability requirement. I suggested that the old `insertionSort` API is there because there would be some data that need to be stable sorted. It's also for future usefulness (or utility?) for the API.

And no, I don't think tiebreaking by PID is enough. htop 's process table can be sorted with multiple keys, and there will be a need to sort the table multiple times by the user _without explicitly specifying the second sort key_. There can be also tables where there isn't an ID field for tiebreaking.

And there's a technical thing: The `Vector` object in htop stores an array of _pointers_ rather than an array of object data themselves. This mean we cannot pass the `Object_compare` function pointers directly to qsort(), but need to make a wrapper for compare function. That's why the third argument of the compare function is needed (and qsort() will not be the best API for the task). There's qsort_r(), but checking the availability of it in configure is also a pain.

Therefore the most future-proof way is to implement a stable sorting function by ourselves. Note that I've developed this algorithm for about half a month, and it would be sad for me if it's ditched finally.

Explorer09 commented:
> If a different algorithm significantly outperforms qsort for sorting processes in htop, we should use it.

@natoscott @BenBE 

Just a minor note. Sorting algorithms' performance can be compared with multiple kinds of data (use cases), and so it won't be a single metric that could decide which algorithm would win.

What I am saying is there can be uniformly random data as well as partially sorted data that need to be sorted. And for an algorithm like heapsort (as an example) it would work best with random data but relatively poor with data that's half sorted. Quicksort with a naïve pivot selection can easily make it the worst case (O(n²)) for data that's already sorted or reverse-sorted, you know. I'm not criticising either algorithm or qsort() API, just a note that the qsort() default algorithm is _not adaptive_ in most libc implementations, thus not as "universally useful" as you think.

And there is no "perfect" sorting algorithm. Each algorithm comes with a trade-off. Even my algorithm does. If you are looking for [block sort family](https://youtube.com/watch?v=InGeRuRk3f8) that looks "perfect" enough in the stats, I did realise there is a tradeoff (adaptive sort i.e. O(n) best case time complexity or O(1) space complexity, pick one). So really no algorithm is perfect.

Explorer09 commented:
> There are differences in the calling conventions between platforms, but they are trivially resolved via wrapper.

The catch is how to detect the calling conventions.

```c
// Older BSD
void qsort_r(void*, size_t, size_t, void*, int(*)(void*, const void*, const void*));
// POSIX, GNU
void qsort_r(void*, size_t, size_t, int(*)(const void*, const void*, void*), void*);
```

Arguments 4 and 5 are both pointers, so only the const qualifier can tell which argument does what. At this point I am no longer interested in writing a `configure` code for this.

natoscott commented:
| The catch is how to detect the calling conventions.

Each platform will know the conventions of its own qsort_r, and a xxx/Platform.h wrapper can map that to whichever convention is chosen for use throughout htop.  The only requirement for configure is to detect whether there is a qsort_r at all ... which AIUI you said some platforms are missing?  If none are missing it, then no configure changes are needed.  If it is missing somewhere, configure would need to enable a generic/qsort.c implementation (with contents from somewhere like https://github.com/freebsd/freebsd-src/blob/main/lib/libc/stdlib/qsort.c for example).

Explorer09 committed:
Make Object_Compare functions accept third argument

Update the prototypes of all Object_Compare functions and Vector_*Sort()
functions to accept the third argument.

The third argument allows passing in extra information or states that
could be useful in sorting. The definition of Object_Compare now matches
the prototype of the compare function in libc qsort_r().

(Previously many programmers would use global variables to read or store
extra information needed in a compare operation. This made the sort
function not reentrant. It was the primary motivation for the qsort_r()
function.)

Currently all Object_Compare functions in htop would simply the third
argument. No functions are using it yet, but this may be changed in the
future.

The callers of Vector_*Sort() will now by default pass in the reference
of the container object (usually named "this") as the third argument to
the sort compare functions.

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
New Vector_sort function, replacing insertion and quick sort

Introduce the new Vector_sort() function and obsolete the old
Vector_quickSort() and Vector_insertionSort() APIs.

This new sort function is a natural, in-place merge sort. I.e. it takes
advantage of partially sorted data, and it's stable.

Space complexity: O(log(n)) worst case (on stack, no malloc())
Time complexity: O(n) best case, O(n*(log(n))^2) average & worst case

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>
